mesa/experimental/cell_space.py

```diff
@@ -0,0 +1,204 @@
+import itertools
+import random
+from collections.abc import Iterable
+from functools import cache, cached_property
+from typing import Callable, Optional
+
+from .. import Agent
+
+Coordinate = tuple[int, int]
+
+
+class Cell:
+    __slots__ = ["coordinate", "_connections", "agents", "capacity", "properties"]
+
+    def __init__(self, i: int, j: int, capacity: int | None = 1) -> None:
+        self.coordinate = (i, j)
+        self._connections: list[Cell] = []
+        self.agents: list[Agent] = []
+        self.capacity = capacity
+        self.properties: dict[str, object] = {}
+
+    def connect(self, other) -> None:
+        """Connects this cell to another cell."""
+        self._connections.append(other)
+
+    def disconnect(self, other) -> None:
+        """Disconnects this cell from another cell."""
+        self._connections.remove(other)
+
+    def add_agent(self, agent: Agent) -> None:
+        """Adds an agent to the cell."""
+        if self.capacity and len(self.agents) >= self.capacity:
+            raise Exception("ERROR: Cell is full")
+        if isinstance(agent.cell, Cell):
+            agent.cell.remove_agent(agent)
+        self.agents.append(agent)
+        agent.cell = self
+
+    def remove_agent(self, agent: Agent) -> None:
+        """Removes an agent from the cell."""
+        self.agents.remove(agent)
+        agent.cell = None
+
+    @property
+    def is_empty(self) -> bool:
+        """Returns a bool of the contents of a cell."""
+        return len(self.agents) == 0
+
+    @property
+    def is_full(self) -> bool:
+        """Returns a bool of the contents of a cell."""
+        return len(self.agents) == self.capacity
+
+    def __repr__(self):
+        return f"Cell({self.coordinate}, {self.agents})"
+
+    @cache
```

Comment:
I tried `@cache` as a decorator of `get_cell_list_contents` in #1995. It gave the error message that looks like https://github.com/projectmesa/mesa/pull/1995#issuecomment-1907405611. Decorators seem to be no-go in Cython.

Reply:
Are you sure? Can’t imagine such a large Python feature not being allowed.

Reply:
I have looked into this somewhere last week. It should be possible to make it work, but will look at this again later.

Reply:
its actually trivial. A decorator is just syntactic sugar. So just do

```python
def some_function():
    ....
some_function = cache(some_function)

```

Which is identical to what the decorator does anyway.

Reply:
I wonder if the method used in #1942 to ensure that the class is pickle-able
```python
def __init__(...):
   ...
   self._original_sum_function = self.some_function
   self.some_function = cache(self._original_some_function)
```
still works.